cache/cache.go

```diff
@@ -1,77 +1,167 @@
+				return err
+			}
+		}
+		return nil
+	})
 	if err != nil {
-		return r, xerrors.Errorf("failed to create cache file: %w", err)
+		return FSCache{}, err
 	}
 
-	tee := io.TeeReader(r, cacheFile)
-	return tee, nil
+	return FSCache{
+		db:        db,
+		directory: dir,
+	}, nil
 }
 
-func (fs FSCache) SetBytes(key string, b []byte) error {
-	filePath := filepath.Join(fs.directory, replacer.Replace(key))
-	if err := os.MkdirAll(fs.directory, os.ModePerm); err != nil {
-		return xerrors.Errorf("failed to mkdir all: %w", err)
-	}
-	cacheFile, err := os.Create(filePath)
+func (fs FSCache) GetLayer(layerID string) []byte {
+	var b []byte
+	_ = fs.db.View(func(tx *bolt.Tx) error {
+		// get a decompressed layer ID
+		decompressedBucket := tx.Bucket([]byte(decompressedDigestBucket))
+		d := decompressedBucket.Get([]byte(layerID))
+		if d != nil {
+			layerID = string(d)
+		}
+
+		bucket := tx.Bucket([]byte(layerBucket))
+		b = bucket.Get([]byte(layerID))
+		return nil
+	})
+	return b
+}
+
+func (fs FSCache) PutLayer(layerID, decompressedLayerID string, layerInfo types.LayerInfo) error {
+	b, err := json.Marshal(layerInfo)
 	if err != nil {
-		return xerrors.Errorf("failed to create cache file: %w", err)
+		return err
 	}
+	err = fs.db.Update(func(tx *bolt.Tx) error {
+		// store a mapping from a layer ID to a decompressed layer ID.
+		if layerID != decompressedLayerID {
+			decompressedBucket := tx.Bucket([]byte(decompressedDigestBucket))
+			err := decompressedBucket.Put([]byte(layerID), []byte(decompressedLayerID))
+			if err != nil {
+				return xerrors.Errorf("unable to store a pair of compressed/decompressed layer IDs: %w", err)
+			}
+		}
 
-	if _, err := cacheFile.Write(b); err != nil {
-		return xerrors.Errorf("cache write error: %w", err)
+		layerBucket := tx.Bucket([]byte(layerBucket))
+		err = layerBucket.Put([]byte(decompressedLayerID), b)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
+func (fs FSCache) MissingLayers(layerIDs []string) ([]string, error) {
+	var missingLayerIDs []string
+	err := fs.db.View(func(tx *bolt.Tx) error {
+		decompressedBucket := tx.Bucket([]byte(decompressedDigestBucket))
+		bucket := tx.Bucket([]byte(layerBucket))
+		for _, layerID := range layerIDs {
+			// get a decompressed layer ID
+			d := decompressedBucket.Get([]byte(layerID))
+			if d != nil {
+				layerID = string(d)
+			}
+
+			b := bucket.Get([]byte(layerID))
+			if b == nil {
+				missingLayerIDs = append(missingLayerIDs, layerID)
+				continue
+			}
+
+			// check schema version in JSON
+			var l types.LayerInfo
+			if err := json.Unmarshal(b, &l); err != nil {
+				missingLayerIDs = append(missingLayerIDs, layerID)
+				continue
+			}
+			if l.SchemaVersion != types.LayerJSONSchemaVersion {
+				missingLayerIDs = append(missingLayerIDs, layerID)
+			}
+
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Since we never return anything but nil, this can be removed.

Reply:
I thought the same thing before, but I found db.View itself sometimes returned an error. For example, it is when db is already closed.